distmv/studentst.go

```diff
@@ -0,0 +1,347 @@
+// conditioned on the input evidence, and the success of the operation.
+// The returned Student's T has dimension
+// n - len(observed), where n is the dimension of the original receiver.
+// The dimension order is preserved during conditioning, so if the value
+// of dimension 1 is observed, the returned normal represents dimensions {0, 2, ...}
+// of the original Student's T distribution.
+//
+// ok indicates whether there was a failure during the update. If ok is false
+// the operation failed and dist is not usable.
+// Mathematically this is impossible, but can occur with finite precision arithmetic.
+func (s *StudentsT) ConditionStudentsT(observed []int, values []float64, src *rand.Rand) (dist *StudentsT, ok bool) {
+	if len(observed) == 0 {
+		panic("studentst: no observed value")
+	}
+	if len(observed) != len(values) {
+		panic(badInputLength)
+	}
+
+	for _, v := range observed {
+		if v < 0 || v >= s.dim {
+			panic("studentst: observed value out of bounds")
+		}
+	}
+
+	s.setSigma()
+
+	newNu, newPhi, newK := studentsTConditional(observed, values, s.nu, s.mu, s.sigma)
+	if newPhi == nil {
+		return nil, false
+	}
+
+	return NewStudentsT(newNu, newPhi, newK, src)
+
+}
+
+// studentsTConditional updates a Student's T distribution based on the observed samples
+// (see documentation for the public function). The Gaussian conditional update
+// is treated as a special case when  nu == math.Inf(1).
+func studentsTConditional(observed []int, values []float64, nu float64, mu []float64, sigma mat64.Symmetric) (newNu float64, newMean []float64, newSigma *mat64.SymDense) {
+	dim := len(mu)
+	ob := len(observed)
+
+	unobserved := findUnob(observed, dim)
+
+	unob := len(unobserved)
+	if unob == 0 {
+		panic("stat: all dimensions observed")
+	}
+
+	mu1 := make([]float64, unob)
+	for i, v := range unobserved {
+		mu1[i] = mu[v]
+	}
+	mu2 := make([]float64, ob) // really v - mu2
+	for i, v := range observed {
+		mu2[i] = values[i] - mu[v]
+	}
+
+	var sigma11, sigma22 mat64.SymDense
+	sigma11.SubsetSym(sigma, unobserved)
+	sigma22.SubsetSym(sigma, observed)
+
+	sigma21 := mat64.NewDense(ob, unob, nil)
+	for i, r := range observed {
+		for j, c := range unobserved {
+			v := sigma.At(r, c)
+			sigma21.Set(i, j, v)
+		}
+	}
+
+	var chol mat64.Cholesky
+	ok := chol.Factorize(&sigma22)
+	if !ok {
+		return math.NaN(), nil, nil
+	}
+
+	// Compute mu_1 + sigma_{2,1}^T * sigma_{2,2}^-1 (v - mu_2).
+	v := mat64.NewVector(ob, mu2)
+	var tmp, tmp2 mat64.Vector
+	err := tmp.SolveCholeskyVec(&chol, v)
+	if err != nil {
+		return math.NaN(), nil, nil
+	}
+	tmp2.MulVec(sigma21.T(), &tmp)
+
+	for i := range mu1 {
+		mu1[i] += tmp2.At(i, 0)
+	}
+
+	// Compute sigma_{1,1} - sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}.
+	// TODO(btracey): Should this be a method of SymDense?
+	var tmp3, tmp4 mat64.Dense
+	err = tmp3.SolveCholesky(&chol, sigma21)
+	if err != nil {
+		return math.NaN(), nil, nil
+	}
+	tmp4.Mul(sigma21.T(), &tmp3)
+
```

Comment:
The space here was abit confusing because you started the previous "paragraph" with "Compute 
sigma_{1,1} - sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}" but here you have computed 
tmp4 = sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1} and then the following subtracts it from sigma_{1,1}. 
It would have been alot more clear to me if line 169 had read "Compute tmp4 which is sigma_{2,1}^T * sigma_{2,2}^-1 * sigma_{2,1}." and then here had a comment saying "compute Sigma_{1,1} - tmp4. Could be just me though.